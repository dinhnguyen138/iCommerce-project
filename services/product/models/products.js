'use strict'
const mongoose = require('mongoose')
const Schema = mongoose.Schema

const schema = new Schema({
  name: String,
  categories: [String],
  description: String,
  rating: Number,
  originalprice: Number,
  currentprice: Number,
  brand: String,
  seller: String
}, {
  timestamps: true
})

module.exports = mongoose.model('products', schema)
