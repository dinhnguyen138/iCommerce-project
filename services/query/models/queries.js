"use strict";
const mongoose = require('mongoose');
const Schema = mongoose.Schema;

const schema = new Schema({
  queryString: String,
  categories: String,
  rating: Number,
  minPrice: Number,
  maxPrice: Number,
  brand: String,
  seller: String,
  sortBy: String,
  sortType: String,
});

// schema.index({'$**': 'text'});
module.exports = mongoose.model('queries', schema);