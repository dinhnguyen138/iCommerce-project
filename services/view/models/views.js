'use strict'
const mongoose = require('mongoose')
const Schema = mongoose.Schema

const schema = new Schema({
  productid: String
}, {
  timestamps: true
})

module.exports = mongoose.model('views', schema)
