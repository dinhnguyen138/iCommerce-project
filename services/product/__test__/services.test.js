"use strict";
const services = require('../services');
const axios = require('axios');
const sinon = require('sinon');
const products = require('../models/products');
const { models } = require('mongoose');
require('../models/products');
require('sinon-mongoose');

jest.mock('axios');
const _data = [
  {
    _id: "0",
    name: "Macbook air 2018",
    categories: ["Laptop", "Electronic", "Apple"],
    description: "Macbook new design 2018",
    rating: 5,
    originalprice: 50000,
    currentprice: 30000,
    brand: "Apple",
    seller: "FPT"
  },
  {
    _id: "1",
    name: "HP Notebook 2020",
    categories: ["Laptop", "Electronic", "Device"],
    description: "Brand new device notebook from HP",
    rating: 4.5,
    originalprice: 45000,
    currentprice: 35000,
    brand: "HP",
    seller: "VienThongA"
  },
  {
    _id: "2",
    name: "Cannon D500",
    categories: ["Camera", "Electronic", "Device"],
    description: "New camera 2020",
    rating: 4.5,
    originalprice: 40000,
    currentprice: 38000,
    brand: "Cannon",
    seller: "FPT"
  }
];
axios.post.mockImplementation( input => {
  return true;
});

let modelsMock = sinon.mock(products);
modelsMock.expects('find').withArgs({ $text: { $search: 'hp' }})
  .resolves([_data[0]]);

modelsMock.expects('find').withArgs({ $text: { $search: 'Cannon' }})
  .chain('exec')
  .resolves([_data[2]]);

modelsMock.expects('find').withArgs({ brand: 'Cannon'})
  .resolves([_data[2]]);

modelsMock.expects('find').withArgs({ categories: 'Electronic'})
  .resolves([_data[1], _data[2]]);

modelsMock.expects('find').withArgs({ rating: { $gte: 4 } })
  .resolves([_data[1], _data[2]]);

modelsMock.expects('find').withArgs({ currentprice: { $gt: 20000, $lt: 30000 } })
  .resolves([_data[0]]);

modelsMock.expects('find').withArgs({})
  .chain('sort', 'brand')
  .resolves([_data[0], _data[2], _data[1]]);

modelsMock.expects('find').withArgs({})
  .chain('sort', '-currentprice')
  .resolves([_data[0], _data[1], _data[2]]);

modelsMock.expects('findById').withArgs(1)
  .resolves(_data[parseInt(1)]);

modelsMock.expects('findById').withArgs(100)
  .resolves(null);

modelsMock.expects('create').withArgs({
    querystring: "camera",
    rating: 4,
    minprice: 22000,
    maxprice: 50000,
    simulate: 'ok'
  })
  .resolves({
    id: "5",
    querystring: "camera",
    rating: 4,
    minprice: 22000,
    maxprice: 50000
  });

modelsMock.expects('create').withArgs({
    querystring: "camera",
    rating: 4,
    minprice: 22000,
    maxprice: 50000,
    simulate: 'failed'
  })
  .resolves(null);

describe('Service test', () => {
  test('getProducts test', async () => {
    let result = await services.getProducts({ query: "hp" });
    expect(result).toMatchSnapshot();
    result = await services.getProducts({ brand: "Cannon"});
    expect(result).toMatchSnapshot();
    result = await services.getProducts({ category: "Electronic" });
    expect(result).toMatchSnapshot();
    result = await services.getProducts({ rating: 4 });
    expect(result).toMatchSnapshot();
    result = await services.getProducts({ price: "20000:30000" });
    expect(result).toMatchSnapshot();
    result = await services.getProducts({ sort: 'brand:asc' });
    expect(result).toMatchSnapshot();
    result = await services.getProducts({ sort: 'price:desc' });
    expect(result).toMatchSnapshot();
  }); 
  test('getProduct test', async () => {
    let result = await services.getProduct(1);
    expect(result).toMatchSnapshot();
    result = await services.getProduct(100);
    expect(result).toMatchSnapshot();
  });
  test('createProduct test', async () => {
    let data = {
      querystring: "camera",
      rating: 4,
      minprice: 22000,
      maxprice: 50000
    };
    data.simulate = "ok";
    let result = await services.createProduct(data);
    expect(result).toMatchSnapshot();
    data.simulate = "failed";
    result = await services.createProduct(data);
    expect(result).toMatchSnapshot();
  });
  test('logQuery test', async () => {
    let data = {
      querystring: 'cannon',
      rating: 4,
      brand: 'Cannon',
      category: 'camera',
      seller: 'fpt',
      price: '20000:30000',
      sort: 'brand:asc'
    };
    await services.logQuery(data);
    expect(axios.post).toHaveBeenCalledTimes(1);
  });
});