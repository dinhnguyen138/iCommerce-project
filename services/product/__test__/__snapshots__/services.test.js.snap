// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service test createProduct test 1`] = `
Object {
  "id": "5",
  "maxprice": 50000,
  "minprice": 22000,
  "querystring": "camera",
  "rating": 4,
}
`;

exports[`Service test createProduct test 2`] = `null`;

exports[`Service test getProduct test 1`] = `
Object {
  "_id": "1",
  "brand": "HP",
  "categories": Array [
    "Laptop",
    "Electronic",
    "Device",
  ],
  "currentprice": 35000,
  "description": "Brand new device notebook from HP",
  "name": "HP Notebook 2020",
  "originalprice": 45000,
  "rating": 4.5,
  "seller": "VienThongA",
}
`;

exports[`Service test getProduct test 2`] = `null`;

exports[`Service test getProducts test 1`] = `
Array [
  Object {
    "_id": "0",
    "brand": "Apple",
    "categories": Array [
      "Laptop",
      "Electronic",
      "Apple",
    ],
    "currentprice": 30000,
    "description": "Macbook new design 2018",
    "name": "Macbook air 2018",
    "originalprice": 50000,
    "rating": 5,
    "seller": "FPT",
  },
]
`;

exports[`Service test getProducts test 2`] = `
Array [
  Object {
    "_id": "2",
    "brand": "Cannon",
    "categories": Array [
      "Camera",
      "Electronic",
      "Device",
    ],
    "currentprice": 38000,
    "description": "New camera 2020",
    "name": "Cannon D500",
    "originalprice": 40000,
    "rating": 4.5,
    "seller": "FPT",
  },
]
`;

exports[`Service test getProducts test 3`] = `
Array [
  Object {
    "_id": "1",
    "brand": "HP",
    "categories": Array [
      "Laptop",
      "Electronic",
      "Device",
    ],
    "currentprice": 35000,
    "description": "Brand new device notebook from HP",
    "name": "HP Notebook 2020",
    "originalprice": 45000,
    "rating": 4.5,
    "seller": "VienThongA",
  },
  Object {
    "_id": "2",
    "brand": "Cannon",
    "categories": Array [
      "Camera",
      "Electronic",
      "Device",
    ],
    "currentprice": 38000,
    "description": "New camera 2020",
    "name": "Cannon D500",
    "originalprice": 40000,
    "rating": 4.5,
    "seller": "FPT",
  },
]
`;

exports[`Service test getProducts test 4`] = `
Array [
  Object {
    "_id": "1",
    "brand": "HP",
    "categories": Array [
      "Laptop",
      "Electronic",
      "Device",
    ],
    "currentprice": 35000,
    "description": "Brand new device notebook from HP",
    "name": "HP Notebook 2020",
    "originalprice": 45000,
    "rating": 4.5,
    "seller": "VienThongA",
  },
  Object {
    "_id": "2",
    "brand": "Cannon",
    "categories": Array [
      "Camera",
      "Electronic",
      "Device",
    ],
    "currentprice": 38000,
    "description": "New camera 2020",
    "name": "Cannon D500",
    "originalprice": 40000,
    "rating": 4.5,
    "seller": "FPT",
  },
]
`;

exports[`Service test getProducts test 5`] = `
Array [
  Object {
    "_id": "0",
    "brand": "Apple",
    "categories": Array [
      "Laptop",
      "Electronic",
      "Apple",
    ],
    "currentprice": 30000,
    "description": "Macbook new design 2018",
    "name": "Macbook air 2018",
    "originalprice": 50000,
    "rating": 5,
    "seller": "FPT",
  },
]
`;

exports[`Service test getProducts test 6`] = `
Array [
  Object {
    "_id": "0",
    "brand": "Apple",
    "categories": Array [
      "Laptop",
      "Electronic",
      "Apple",
    ],
    "currentprice": 30000,
    "description": "Macbook new design 2018",
    "name": "Macbook air 2018",
    "originalprice": 50000,
    "rating": 5,
    "seller": "FPT",
  },
  Object {
    "_id": "2",
    "brand": "Cannon",
    "categories": Array [
      "Camera",
      "Electronic",
      "Device",
    ],
    "currentprice": 38000,
    "description": "New camera 2020",
    "name": "Cannon D500",
    "originalprice": 40000,
    "rating": 4.5,
    "seller": "FPT",
  },
  Object {
    "_id": "1",
    "brand": "HP",
    "categories": Array [
      "Laptop",
      "Electronic",
      "Device",
    ],
    "currentprice": 35000,
    "description": "Brand new device notebook from HP",
    "name": "HP Notebook 2020",
    "originalprice": 45000,
    "rating": 4.5,
    "seller": "VienThongA",
  },
]
`;

exports[`Service test getProducts test 7`] = `
Array [
  Object {
    "_id": "0",
    "brand": "Apple",
    "categories": Array [
      "Laptop",
      "Electronic",
      "Apple",
    ],
    "currentprice": 30000,
    "description": "Macbook new design 2018",
    "name": "Macbook air 2018",
    "originalprice": 50000,
    "rating": 5,
    "seller": "FPT",
  },
  Object {
    "_id": "1",
    "brand": "HP",
    "categories": Array [
      "Laptop",
      "Electronic",
      "Device",
    ],
    "currentprice": 35000,
    "description": "Brand new device notebook from HP",
    "name": "HP Notebook 2020",
    "originalprice": 45000,
    "rating": 4.5,
    "seller": "VienThongA",
  },
  Object {
    "_id": "2",
    "brand": "Cannon",
    "categories": Array [
      "Camera",
      "Electronic",
      "Device",
    ],
    "currentprice": 38000,
    "description": "New camera 2020",
    "name": "Cannon D500",
    "originalprice": 40000,
    "rating": 4.5,
    "seller": "FPT",
  },
]
`;
